### PR TITLE
VideoPress: Add settings to the React interface

### PR DESCRIPTION
**Pull request by rcoll**
#### Changes proposed in this Pull Request:

Bring VideoPress settings over to the React interface. I think this is pretty solid so far but would love @dereksmart's input on the React code, as well as some experienced eyes on the API bits. This PR should be considered in-progress for now. 

screen shot 2016-09-30 at 2 49 08 pm

screen shot 2016-09-30 at 2 51 48 pm
#### Testing instructions:
- Spin up a new Jetpack site and activate VideoPress. Follow the rabbit hold within VideoPress settings to create and connect a new WordPress.com Premium account to activate VideoPress. Ensure you can upload and embed videos as usual. Ensure all VideoPress settings save and reflect properly.
- Apply this PR to an existing VideoPress site and ensure you can upload and embed videos as usual. Ensure all VideoPress settings save and reflect properly.
- Attempt to alter settings and/or use VideoPress features as an Author and ensure that settings prevent unauthorized access where configured to do so.

#### Proposed changelog entry for your changes:
---
- [ ] Make sure your changes respect [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- [ ] Did you make changes, or create a **new .js file**? If [Gulp](http://gulpjs.com/) is installed on your testing environment, run `gulp js:hint` before to commit your changes. It will allow you to [detect errors in Javascript files](http://jshint.com/about/).
- [ ] Did you create a **new action or filter**? [Add inline documentation](https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/php/#4-hooks-actions-and-filters) to help others understand how to use the action or the filter.
- [ ] Create **[unit tests](https://github.com/Automattic/jetpack/tree/master/tests)** if you can. If you're not familiar with Unit Testing, you can check [this tutorial](https://pippinsplugins.com/series/unit-tests-wordpress-plugins/).

### CODE REVIEW

**Comment by samhotchkiss** [comment]
@rcoll and @dbtlr -- We want to eliminate many/all of these settings for new users, replacing them with smart defaults and filters.  @MichaelArestad is going to help get those sorted out, so have a chat with him!

**Comment by MichaelArestad** [comment]
Yep. Kiss goodbye to all VideoPress settings!

Defaults:
* Library access: allow users to access my videos
* Allow users to upload videos
* Free formats: display all formats (not just free ones)
* Quality: Always high quality!

The biggest issues facing us are around connected blogs. It's a confusing concept to begin with.

Connected blogs problems:
* VideoPress requires a connected blog. Can it instead just use the shadow site removing the need for connect blogs?
* Users currently can use one blog to power VideoPress across all of their sites. If we change this behavior or make it a single site thing, it could break blogs. We could let sites using these keep doing that but remove the option. Ideas?
* Limiting it to one site, which I think we should, could upset previous users who use it across blogs.

**Comment by EnigmaSolved** [comment]
> Limiting it to one site, which I think we should, could upset previous users who use it across blogs.

I'd just like to chime in that I'd be rather disappointed to lose the ability to use VideoPress on multiple sites. In my case it is just two sites: my main/live/public site, which exists on one domain, and then a sub-directory of a different domain where I do my testing/staging. I use VideoPress in the mode where it is fully embeded into the page (ie, no Iframe) so that I can style the CSS, and consequently I _**really**_ appreciate being able to update Jetpack on my staging domain first to make sure there's no breaking changes, and then update Jetpack on my public domain. So I'd be pretty disappointed (and rather upset) to lose the ability to use my single VideoPress account across those two domains.

I think that if you take away that functionality it is going to feel like a money-grab to users, as there's no reason (that I can see) to limit things in that way, and it's not how the cost structure has been previously (ie, cost has been built around how much space you're paying for).

That's my 2 cents.

**Comment by dbtlr** [comment]
@EnigmaSolved: I'm not going to go too much into the trolling you just did around cash grabs. This is a pure simplification of an already existing, very confusing issue for 99% of VideoPress users. As of right now, the Jetpack plan is a direct price analog to the wpcom account that they would be paying for. We're not asking them to purchase anything new. Instead, we are making every effort to seamlessly migrate everyone, so there is little to no negative impact. For that, they get the added benefits of the Jetpack plan itself.

In terms of your very specific (and rare) use case, while it is certainly not forbidden to do what you're doing, it is also a behavior that is part of an extreme minority of VideoPress users. You're welcome to continue doing what you're doing. Since you're ignoring the shortcode processor altogether, there is absolutely nothing stopping you from embedding the video directly, just as you do now. This part of the code is not being touched in any way. 

Between this change and #5326, the only things that are changing are more admin UX related. And it is not being done as a cash grab, but as a simplification in an effort to make VideoPress usable.

In the end, the number of users using VideoPress across blogs is incredibly small and it would be far better for us to optimize for the majority of users and perhaps find a way to help these other users on a one-off basis.

**Comment by EnigmaSolved** [comment]
@dbtlr, thanks for responding kindly, especially in light of my own post. :)

I apologize for the negative tone of my post--I should have reflected longer before writing, and written differently.

Just to clarify, I wasn't meaning to suggest that _I_ myself believed this was a cash grab--I am aware it is being driven by important and very helpful changes in the architecture--only that in my experience mass users don't see it that way (they just see a loss of functionality that requires them to pay more to do the same things they've been doing).

Having said that, it sounds like I misunderstood the prior post ( https://github.com/Automattic/jetpack/pull/5233#issuecomment-256513123 ), because I interpreted it to suggest that there's more than a small number of blogs using Jetpack across multiple site (hence my suggesting you're kinda asking for flak if you take away something a lot of folks are doing). I'm sorry for my misread.

I also wanted to clarify and ask a question regarding what I'm doing: I do use the VideoPress shortcode, but I use a built-in VideoPress filter ( https://gist.github.com/EnigmaSolved/5f848b5270fe93b601b68a2736ef7278 ) to disable Iframes so that I can then style the CSS. I'm doing the same exact thing on both my Development domain, and my Live/Public domain (but with different videos displayed on each--I use some test videos on the Develpoment domain, typically). Is that going to become problematic with what you all are proposing?

Thank you again for responding so kindly. I ought to have written my post differently (and/or not written when I was tired). I am sorry. :( It is true that I don't want to lose functionality, and that it would be frustrating for me if that happens. But I did not, and don't, believe the actual motivation on your parts is money, but rather trying to improve things in various ways (which I am very grateful for--I agree with you that the VideoPress integration has been less than ideal). I'm sorry for not being clearer, not asking more questions first, and for "trolling" by putting forth (in a messy, unclear manner) my interpretation of how the internet might react to what I was perceiving to be the case. You took the higher road in your reply, and I am very, very appreciative. :)

**Comment by dbtlr** [comment]
@EnigmaSolved: No problem. Glad I could help shed some light on things. Very cool to see how you're overriding things via that hook. Very few people are doing that.

The shortcode itself is available with or without the VideoPress module being enabled, which means your code will keep working just as it is, whether-or-not you have a Jetpack plan. This means, if you have one Jetpack blog with a plan, you can upload all your videos to that and simply include the shortcodes on any Jetpack-enabled blog, regardless of whether you've paid for a plan on that blog. So, pretty much the same as it is now. :)

In terms of the prior comment about the blogs using it for multiple sites. This wasn't something that we had hard data on until recently, so it was a little conjecture and a lot of due diligence that brought up that question. More than anything, we want to make sure we discuss handling every user as best as we can. With a move like this, we believe it is what is best for not only the product but for the ecosystem as a whole. To that end, we're doing our best to ensure that we take everyone into account and make sure their experience keeps working as best as we can.

In the long run, our goal is to make VideoPress and the rest of Jetpack as dead simple as possible. Since one of the biggest confusions with it has always been around creating a WordPress.com blog to host the videos, we want to clean up that mess and make sure that we make the experience as seamless as possible.

**Comment by EnigmaSolved** [comment]
@dbtlr, thanks for understanding, and for your compliment. :) Yeah, I think I only learned about that hook by emailing support to find out how I could apply my own CSS to the player (since there is no way to do that when it is loaded via Iframe). I know that feature (of customizing the CSS) isn't being pushed right now (it's considered Beta), but it (in conjunction with being able to fully embed the video player sans-iframe) is one the things I really like about VideoPress. You can see on my [website](https://transitionsmft.org/blog/822/facing-unfaceable-reflections-on-challenging-work-personally-professionally-love-talk-expert-panel-2015/#goto-v-7UboJ61j-1) how I've been able to significantly alter the look of the player.

On a related note, it would be great to eventually have the player CSS classes cleaned up a bit. Currently they appear to inject a (seemingly) random, unique string into all classnames for each player, presumably to allow for targeting CSS on an individual player level. But it makes writing CSS for styling _all_ players rather cumbersome, as you have to use selectors like `[class*="_share-inner"] {}`. It's not a huge deal, but would be nice to have a better system (eg, maybe a root element classname combo that has one class that is unique to that player, and then one that is shared by all players, so that we can more easily create CSS descendant selectors off of that for either scenario).

> The shortcode itself is available with or without the VideoPress module being enabled, which means your code will keep working just as it is, whether-or-not you have a Jetpack plan. This means, if you have one Jetpack blog with a plan, you can upload all your videos to that and simply include the shortcodes on any Jetpack-enabled blog, regardless of whether you've paid for a plan on that blog. So, pretty much the same as it is now. :)

That's brilliant, and so obvious in retrospect! Thanks for pointing that out. :) As was apparent in my earlier post I was wrongly thinking I'd lose the ability to achieve that. But you're absolutely right that in my case I don't need to be able to upload from multiple blogs--in fact, I currently upload directly via the wordpress.com interface because that worked better with my browser back when I started using VideoPress (and I realize that is one of several things you all are working on improving--the enabling of more intuitive uploading from within Jetpack).

I do agree with the direction you all are heading. For average users it is probably quite confusing how things have functioned in the past (ie, creating a wordpress.com blog, etc.). I would imagine/hope that as you implement these changes of simplifying things that you'll see an increase in adoption (and possibly an increase in interest in lesser known, but very useful, features like the CSS customization).

Thanks again for clearing things up and bearing with me. :)

**Comment by samhotchkiss** [comment]
#5457

**Comment by MichaelArestad** [comment]
@dbtlr What are the next steps here? Can we resolve this issue or is there a need for a setting or two, still?

**Comment by rcoll** [comment]
I think the plan is to do away with the options, default to smart settings, and use the connected Jetpack plan for authorization instead of having the dependency on a wpcom blog. I'm assuming this particular interface will show "no options" when a plan is in use, or have an upsell if there's no Jetpack plan.

I'll wait on @dbtlr's thoughts before I do anything.

**Comment by dbtlr** [comment]
Correct, we need to have something in place, similar to VaultPress, that
tells the user to purchase a plan to turn VideoPress on. Otherwise, I love
what you guys have so far. I'll set the defaults in the backend and I'd
love to see as many of these extra fields just go away if possible. :)

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 1]
```diff
@@ -391,6 +391,41 @@ public function update_module( $data ) {
...
+					break;
+
+				case 'videopress_library_access':
+					$vp_options = Jetpack_VideoPress::init()->get_options();
+					$vp_options['access'] = sanitize_text_field( $value );
+					Jetpack_VideoPress::init()->update_options( $vp_options );
+					$updated = true;
+					break;
+
+				case 'videopress_allow_user_upload':
```
Comment (dbtlr) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Removed

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 2]
```diff
@@ -391,6 +391,41 @@ public function update_module( $data ) {
 			$value = Jetpack_Core_Json_Api_Endpoints::cast_value( $value, $options[ $option ] );
 
 			switch ( $option ) {
+				case 'videopress_connected_blog':
```
Comment (dbtlr) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Removed

**Pull request closed without merging**